### PR TITLE
Note hover tooltip

### PR DESCRIPTION
**Pull request by Derkness**
- [Need Guidance. Where to put this?] I documented all behaviour as far as I could do.
  - [Yes] I target the develop-branch, and *not* the master branch.
  - [Yes] I have tested this extensively and paid extra attention to
        potential cross-platform issues (e.g. Cmd/Ctrl/Super-key bindings)
  - [Two warnings I can't think of how to remove ] I have made use of ESLint using the provided configuration from the
        repository's .eslintrc.json file, and it did not complain.
  - [Unsure where to put this in the file] I have added an entry to the CHANGELOG.md.
  - [I believe so] I matched my code-style to the repository (as far as possible).
  - [Yes] I do agree that my code will be published under the GNU GPL v3 license.
  - [Yes] As far as JS-files are concerned, I made sure to copy (in case of new files)
        or adapt (in case of existing files) the info in the header.
  - [Not fully sure what this means] I synced the latest commits to develop shortly before proposing
        so that no merge issues occur.

## Description

Hovering over an internal note link will display a tooltip. This contains: The file name, the first 200 charactrs, the word count, and the time since modification (in days, or hours, or minutes, choosing the largest digit greater than 0)

## Changes

I added a command, and linked it through index.ts

## Additional information

This is based on issue #920 (https://github.com/Zettlr/Zettlr/issues/920)
Testing:
- emojis (works perfectly fine)
- tabs and spaces are reduced to a single space to reduce clutter
- lines of code, test with <title>test</title> have the html tags removed
- bold and Italics shows the ** and _ just like the note doc
- dot point lists (shows the * but of course shows inline. This is fine)
- tables show as text seperated by '|'
- day/hour/minute works as expected, and each threshold is passed to the next 'tier' up
Concern:
- code comments seem to not be visible

Tested on: Windows 10 (current build)

### CODE REVIEW

**Comment by kyaso** [comment]
I actually started working on something similar, what a coincidence! Probably shows that a tooltip for ZKN links can be a very handy feature.

If I understood your code correctly, a crucial feature (in my eyes) is missing: being able to click on the link to open the target file (+ global search, if enabled in options). That's actually what I was planning to implement. However, I haven't really started with that yet, so I unfortunately can't give you hints on that right now..

Also, does your approach work for ID-based links as well? Eg, [[20210520101431]].

**Comment by kyaso** [comment]
Maybe I should also mention what I had in mind.

I was just going to implement a simple tooltip that just shows the ZKN link content. When you click on it, the same steps happen as if you CTRL/Cmd + clicked the link in the classical way (that is: open target file (if present), and do global search).

So it would work any kind of ZKN link. Eg, [[20210520101431]] (ID-based), [[my_file]] (file name based), or even just a link which acts as shortcut for global search [[search term(s)]].

Btw, there is also an issue #1454 already that asked for ZKN tooltips (mentioning here for reference)

**Comment by Derkness** [comment]
Thanks for your review!
The issue we were tasked with solving (#920) was just to hover over a link and get a tooltip. Since the link can be ctrl+clicked normally with pre-existing systems, that wasn't included in the tooltip. However, I can add that feature if that is the general consensus of this community.

WIth the ID aspect, I will try to find a way to add that aswell. It currently does not work, thanks for spotting that :D

I'm not quite sure what you mean with respect to a global search though. The issue linked is just about seeing a preview of the document's information. Unless you mean searching for the file itself, which this does do I believe.

**Comment by kyaso** [comment]
Ah, I wasn't aware of #920.

With global search I meant the search that gets initiated whenever you CTRL+click a ZKN link.

I believe, the current `develop` branch doesn't seem to trigger a global search on CTRL+click anymore (#1989), but it works in 1.8.x.

**Comment by Derkness** [comment]
Tooltip previews for ID based ZKN links is now implemented, thanks for the spot Kyaso!

**Comment by Derkness** [comment]
Furthermore, surrounding having a button in the tooltip to link. Would that be something the community desires? It seems redundant to me since clicking the link will already redirect to the note, but I can duplicate this if desired :D

**Comment by kyaso** [comment]
> Furthermore, surrounding having a button in the tooltip to link. Would that be something the community desires? It seems redundant to me since clicking the link will already redirect to the note, but I can duplicate this if desired :D

I personally would love to see a clickable link at the bottom of the tooltip. That way, one doesn't have to do CTRL+click everytime, and navigating becomes much faster.

We already have that feature for external ("normal") links, so having it for ZKN links also, makes sense, IMO.

Let's see what @nathanlesage thinks of this (and the overall PR, of course).

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:25]: Please add a comment `// TODO: Translate!` after any one of these lines, just so that I can perform a big, broad search for all of the TODOs shortly before releasing 2.0 so I don't forget to translate those strings (the important part is the all-uppercase "TODO" keyword)

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:26]: btw wouldn't it make sense to add an ellipsis here to indicate a process …? Or, even better, a loading spinner?

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:29]: The tippy below won't display the loading values, no, it'll just display "Searching For File ...". Did you mean to say "This will contain the values **after they have been loaded**"?

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:39]: Actually, it might make sense not to destroy these tippys. I'm not sure about the performance impact of leaving the tippys, BUT I think this way we could save loading time if the user accidentally leaves the element and immediately hovers back over it (because if there's a `_tippy` associated with the link, then tippy will itself react to mousemove events and thus, re-display the tippy).

Would you mind taking a look into the Tippy documentation and maybe fiddle around a little bit with leaving the tippy vs. not leaving them? And also think about the potential usability impact if the notes get edited. If we leave the tippy instance and the user never scrolls, the loaded values might become outdated, the question would only be if that is ever a problem… I think not because as soon as the user switches documents (to edit the linked one) this `event.target` would get destroyed either way alongside the tippy and would be re-rendered either way.

So long story short, you may want to play around a little bit with some settings and tippy! You can also take a look at the formatting toolbar (it's either a hook or a plugin of the editor) which I implemented, since that one features some fancy tricks with the Tippy instance.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:45]: I would refactor so that the callback just calls a different function with the loaded file contents. Plus I just saw you're invoking two asynchronous calls in succession, and that's not only callback hell but a hell to maintain. I don't have an ad-hoc solution, but the good solution will definitely use functions and a better state management. Mh … we have to get to this.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:47]: Don't do this. Just don't. That's what `undefined` or `null` are for. Or `false` (although in this case it MIGHT make sense to better use `undefined`. I'll comment that on the main processes' command once I get there in a few minutes)

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:52]: Don't do this. I've recently added an HTML sanitizer that works fine for this. In general, I'd suggest we display a rendered HTML source (a.k.a. run the whole content through the `md2html`-source). But we might want to take a look at that, since something like this got me a CVE a few days ago, so … remove that line for now, but we must talk about this.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:65]: … what are you doing here? This looks both unnecessarily tedious, has to many nested `if`-statements and I'm not sure why we'd need this in the first place …?

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:71]: In case you mean "title" in the general Zettlr one then you need to implement the triage of horror I added in other places, where first of all a YAML frontmatter title is being used, THEN the first h1 heading, and only THEN the filename (but without extension in most places), depending on certain configuration options.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:74]: Erm, for this we have `luxon` bundled. You don't have to manually (+ non-localised) implement that thing. Take a look at the file list item Vue component, there I'm using that to precisely display a time difference.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:96]: Add a TODO translate comment here as well, please

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:94]: There's a grammatical error in here. Please rephrase to "Could not load file for preview: "

File: source/main/commands/file-path-find.ts
Comment (nathanlesage) [inline source/main/commands/file-path-find.ts:42]: You know what I just realised? Up in the editor hook you are running two successive asynchronous IPC calls to first retrieve the file path and THEN the note's contents. However, why don't you combine these calls, and instead of `return file.path` use `return await this._app.getFileSystem().getFileContents(file)`? Then you have everything you need, skip the second IPC call altogether and can safely refactor the code in the callback into its own function.

File: source/main/commands/file-path-find.ts
Comment (nathanlesage) [inline source/main/commands/file-path-find.ts:40]: Are you sure `findExact` returns undefined if it doesn't find anything? (If yes, then please remind me to change that to `null`, that would be an error on my side)

File: source/main/commands/file-path-find.ts
Comment (nathanlesage) [inline source/main/commands/file-path-find.ts:46]: Here's where you should return `null`, `undefined`, or `false`, but not a "Not Found" string.

File: source/main/commands/file-path-find.ts
Comment (nathanlesage) [inline source/main/commands/file-path-find.ts:50]: Remove this line, that's the old NodeJS module system, and you already have `export default class` above, as it's supposed to be.

Plus, I think it might sound nicer if you had it named `FileFindPath` (or, with my changes, a different, more fitting name) – don't feel _too_ obliged to copy my "But File/Dir has to go first!!"-approach if that makes it sound like complete nonsense …

File: CONTRIBUTING.md
Comment (Derkness) [reply CONTRIBUTING.md:17]: In the conversation from (#920) it was discussed I could add a quick guide on how I got the development setup working. I also added a bit of text, was my format done badly?

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (Derkness) [reply source/common/modules/markdown-editor/hooks/note-preview.js:29]: Sorry. The intented functionality is to just display "Searching For File...". the fact the values are initialised to 'loading' is a remnant form a previous build plan. I can set them to initialise to 'null'

File: CONTRIBUTING.md
Comment (nathanlesage) [reply CONTRIBUTING.md:17]: Aaaah that's where that comes from. Please remove these changes from this PR, open a new PR for that and let's keep the PRs separate.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (Derkness) [reply source/common/modules/markdown-editor/hooks/note-preview.js:65]: This was because if I just assumed there was enough content and added newlines, I would get an extra space at the end oof the content. I will refactor this so it's a loop with just a scaling conditional check each time though.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (Derkness) [reply source/common/modules/markdown-editor/hooks/note-preview.js:71]: in this case it's just the filename, to properly identify the file being referenced (mostly useful in the case where an ID was used, but also sueful in order to 'ground' yourself when looking at a large tooltip)

File: source/main/commands/file-path-find.ts
Comment (Derkness) [reply source/main/commands/file-path-find.ts:40]: I believe so. I initially set this check to 'null', however it was being triggered when there was no file as 'null' was the response of findExact.

File: source/main/commands/file-path-find.ts
Comment (Derkness) [reply source/main/commands/file-path-find.ts:42]: I agree. I will change the 'file-path-find' to achieve this. Probably renaming it for its new functionality. Would you like me to keep an old version of the file-path-find around so that other people have a command to just grab the absolute path of a file?

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (Derkness) [reply source/common/modules/markdown-editor/hooks/note-preview.js]: Would you have any tips on where to put the gif for this spinner?

Comment (nathanlesage) [reply source/common/modules/markdown-editor/hooks/note-preview.js]: I don't know if we should use a GIF – we could do this readily with SVG or even JavaScript-animated HTML. After all, we do have the space, and a single, continuous animation shouldn't be too demanding (but the quality improvements are massive).

One quick Google search yielded, e.g., this one: https://loading.io/css/

And I remember I had a loading spinner animation in an older version of Zettlr, maybe we could recycle that one by having a look into the old versions, it should be ~somewhere~ there!

File: source/main/commands/file-path-find.ts
Comment (nathanlesage) [reply source/main/commands/file-path-find.ts:42]: Nah, I don't think we should do that. I've learned the hard way that you should only implement something as soon as you need it, and not in good faith that maybe someone uses it some day, so it's safe to just overwrite that command! (After all, it's super simple to implement that if we really need it)

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (Derkness) [reply source/common/modules/markdown-editor/hooks/note-preview.js]: Sure, I'll put one of those in. Would you be able to tell me where to put the css for the note-preview.js file I made?

Comment (nathanlesage) [reply source/common/modules/markdown-editor/hooks/note-preview.js]: hm... "Somewhere reasonable", I don't have the source code in front of me, so I'll leave the decision to you for now. I mean moving around a file shouldn't be a deal breaker! :)

Comment (Derkness) [reply source/common/modules/markdown-editor/hooks/note-preview.js]: I am a bit stuck with trying to get my css into my javascript. From my research, I can simply import the document. However this file must use the 'require' method, and even then I cannot find the file.

My chosen location for the file is a in a folder called css in the hooks folder. Therefore, the following code should work:

const { tippyCSS } = require('css/note-preview.css')

right? Let me know if I am way off here, im not quite sure with this stuff.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (Derkness) [reply source/common/modules/markdown-editor/hooks/note-preview.js:40]: It seems that leaving the tippy up does not prevent the content being changed. Probably because, as you said, when navigating away to change the tippy is removed anyway. Because of this, I have removed the onHidden event all together.

I think that this means that when a note is edited in a second of Zettlr, then the tooltip in the first instance won't be updated, but that doesn't seem like a massive deal compared to the cost of re-creating each hover time.

**Comment by Derkness** [comment]
Side note: As I was ticking off all the changes in my head, I was marking them as resolved to collapse them. I thought this was local to just me, but I think that the standard approach is you are meant to mark them as resolved when you're happy with the changes. So I apologise if I distrupted the standard there.

**Comment by Derkness** [comment]
Hey guys, sorry to bother but I'm just wondering if I could get another review? I believe I have made the changs requested. The only thing remaning is the possible loading spinner, but I have posted a question about that in the discussion for it.

I only ask for an update as this is for a Uni Assignment, and it is due in 3 days.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:48]: I still do not understand why you go through the hassle and inserting linebreaks every 50 characters here. Why not take a substring and then at least provide an algorithm for finding the next space accounting for possible Markdown markup?

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:71]: Alright, you're using Luxon, but you're still doing weird stuff. Point taken, the Luxon docs are hard, but have you tried this function (or something along these lines)?

https://github.com/Zettlr/Zettlr/blob/develop/source/common/util/format-date.js

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:42]: It would still be great if you could move all that logic to a separate function which you call here (either by simply passing it as a reference, or by calling it from within the callback).

**Review by nathanlesage — CHANGES_REQUESTED** [review]
It's coming together.

One thing: I literally have no time and also to prepare for an exam on June 1st, so if you notice someone else waiting for a PR review, **please explain to your instructor that I cannot guarantee reviews by May 31st, but will do so asap afterwards**. This way people whose notifications are drowning in the 60 I have still open here won't get any problems with examination.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (Derkness) [reply source/common/modules/markdown-editor/hooks/note-preview.js:71]: I did try to do that, but for some reason I can't get the function into my code. Not sure what im doing wrong, as all the found usages use the 'import' method, whereas I have to use the require method. I am not sure why this like isnt working:

const { formatDate } = require('../../../util/format-date.js')

and then running it as:

formatDate(dateDif)

I have tried to research this, but I can't find the words to describe exactly what I'm doing.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (Derkness) [reply source/common/modules/markdown-editor/hooks/note-preview.js:48]: I honestly just assumed I had to manually set newlines. I didn't realise tippy autoformatted.

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [reply source/common/modules/markdown-editor/hooks/note-preview.js:71]: You should be able to use `import { formatDate } from '../../../util/format-date'`

Requires are deprecated either way.

Additionally, you need to pass a single Date object (or milliseconds; you can pass `descriptorWithContent.modtime` to that function), because `Date.now()` will automatically be instantiated within the function. You do _not_ need to pass a timedelta.

**Comment by Derkness** [comment]
I have made the changes. Specifically I have turned find-file-and-return into find-file-and-return-meta-data. I figued this would be better because it makes the function more specific, and you mentioned before leaving loose programming things around for others to maybe use is a bad idea.

**Comment by Derkness** [comment]
This was working when I submitted it, but I changed branches to contribute to #2042, and then on switch back everything is broken. I assume this is just something to do with branch switching considering the pushed versino works, but please let me know.

**Comment by nathanlesage** [comment]
> Additionally, my instructor has allowed for our presentation to go ahead without the completion of this, however I will still stick around to contribute anyway.

That is very nice and I appreciate the spirit! I'm right now dedicated to code-reviewing all your PRs!

File: source/main/commands/file-find-and-return-meta-data.ts
Comment (nathanlesage) [inline source/main/commands/file-find-and-return-meta-data.ts:34]: Please use camelCase for this variable

File: source/main/commands/file-find-and-return-meta-data.ts
Comment (nathanlesage) [inline source/main/commands/file-find-and-return-meta-data.ts:46]: You should break here to save some resources

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:47]: It might make sense to just log the error here (the reason is that this way we'll have the stack trace)

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (nathanlesage) [inline source/common/modules/markdown-editor/hooks/note-preview.js:34]: Wouldn't an arrow pointing to the link contents make sense …?

**Review by nathanlesage — COMMENTED** [review]
Sooo, here's finally the promised review! In general it looks pretty good; there are a few æsthetic things I've commented, but in general it looks good. I'll be testing this over the next week and then I think it can be merged!

File: source/common/modules/markdown-editor/hooks/note-preview.js
Comment (Derkness) [reply source/common/modules/markdown-editor/hooks/note-preview.js:34]: I am happy to set the arrow to true, however I should warn that that is inconsistent with the current hyperlink tooltips.

Comment (nathanlesage) [reply source/common/modules/markdown-editor/hooks/note-preview.js]: Yes, the hyperlink tooltips don't use that. But the formatting toolbar and the footnotes tooltip use it. I think we need to focus on the question: "Should it point to something?" and I think for those note tooltips it makes sense to point back to the link!

**Comment by Derkness** [comment]
Awesome! Those final changes have been completed and everything should be done :D (i think :P)

**Comment by nathanlesage** [comment]
Works like a charm! Merging …

**Comment by Derkness** [comment]
Awesome! I just wanted to thank you on supporting me throughout this project. I don't think I would have been able to do it without your help, and it was a very good experience to learn how open source worked. I know when you started this project you didn't know it would be the basis of a Uni assignment, but the fact this project has had that effect is quite cool!

**Comment by nathanlesage** [comment]
Absolutely! It was a pleasure! And of course, feel free to continue contributing whenever you have time!

**Comment by nathanlesage** [comment]
Just wanted to let you know that I've just attempted at vastly improving the display of these tooltips:

image

Hope you like it! It's in https://github.com/Zettlr/Zettlr/commit/8266492c9a5e8a56b2985eb944ec9958e6ef302e

**Comment by Derkness** [comment]
Awesome! That looks a lot nicer! I'm glad my backend was still usable after the redesign.

**Pull request merged**
